### ISSUE TITLE
Add Cookies in playwright.config.ts based on the environment variable 

### ISSUE THREAD

**Issue by dipesh17** [microsoft/playwright#22431]
I have a requirement as follows;
baseURL : ‘[www.microsoft.com](http://www.microsoft.com/)’

I want to execute my test suite on the test environment as well as the production environment.

To switch the baseURL to the environment I need to add a header
‘User-Agent’ : ‘TestSegment’

This will point the application to the test environment.
This i have achieved through extraHTTPHeader property in the playwright.config.ts

When I tried to add cookies to the test suite using the storageState property in the playwright.config.ts I am not able to add cookies to the test suite.

![Screenshot 2023-04-17 at 13 16 58](https://user-images.githubusercontent.com/26460424/232484432-fb52114a-1dc8-4713-af2e-f66589e22e4b.png)

In the above example, i am expecting cookies banner on the home page to disappear however it does not.

In a microservice’s environment, when we deploy all respective services in a particular segment to test on the test environment to validate if the application work on the test environment using the segment. 

I would like to add cookies to the baseURL if the environment is a test environment.

CC_TEST_ENV=TEST SEGMENT='DIPESH' npx playwright test

This means I am pointing my test suite to the Test environment by the environment variable CC_TEST_ENV=TEST and adding cookies to the browser segment: “dipesh”

The condition is that we should add segment  cookies only if we  CC_TEST_ENV=TEST

I am unable to do this, can someone please help me here.

**Comment by dipesh17** [comment]
const addSegmentCookies = async () => {
    if (process.env.CK_ENV?.toUpperCase() === 'TEST' && 
        process.env.SEGMENT != undefined && 
        process.env.SEGMENT?.length> 0){
            const context = request.newContext({
                baseURL:"https://envs.corp.ms.co.uk"
            })
            const response = await context.get("/segments/" + process.env.SEGMENT + "?format=token")
            const driToken: string = response.json()
           return {
                name: 'segment',
                value: driToken,
                path: '/',
                domain: ccBaseURL,
                expires: 99999999,
                httpOnly: false,
                secure: false,
                sameSite: 'None'
            }
    }
    return undefined;
}


/**
 * See https://playwright.dev/docs/test-configuration.
 */
export default defineConfig({
    testDir: './tests',
    /* Maximum time one test can run for. */
    timeout: 30 * 1000,
    expect: {
        /**
         * Maximum time expect() should wait for the condition to be met.
         * For example in `await expect(locator).toHaveText();`
         */
        timeout: 5000
    },
    /* Run tests in files in parallel */
    fullyParallel: true,
    /* Fail the build on CI if you accidentally left test.only in the source code. */
    forbidOnly: Boolean(process.env.CI),
    /* Retry on CI only */
    retries: process.env.CI ? 2 : 0,
    /* Opt out of parallel tests on CI. */
    workers: process.env.CI ? 1 : undefined,
    /* Reporter to use. See https://playwright.dev/docs/test-reporters */
    reporter: [['junit', { outputFile: 'circuit_output/results.xml' }]],
    /* Shared settings for all the projects below. See https://playwright.dev/docs/api/class-testoptions. */
    use: {
        /* Maximum time each action such as `click()` can take. Defaults to 0 (no limit). */
        actionTimeout: 0,
        /* Base URL to use in actions like `await page.goto('/')`. */
        baseURL: ccBaseURL,

        /* Collect trace when retrying the failed test. See https://playwright.dev/docs/trace-viewer */
        trace: 'retain-on-failure',
        browserName: 'chromium',
        headless: true,
        extraHTTPHeaders: {
            ...getUserAgent(),
            ...getFEHeader()
        },
        storageState: {
            origins: [],
            cookies: [
                {
                    name: 'cc_cookie_accept',
                    value: 'cc_cookie_accept',
                    path: '/',
                    domain: ccBaseURL,
                    expires: 99999999,
                    httpOnly: false,
                    secure: false,
                    sameSite: 'None'
                },
                ...addSegmentCookies()
            ]
        },
        bypassCSP: true,
        launchOptions: {
            args: ['--disable-web-security']
        }
    }

    /* Folder for test artifacts such as screenshots, videos, traces, etc. */
    // outputDir: 'test-results/',

    /* Run your local dev server before starting the tests */
    // webServer: {
    //   command: 'npm run start',
    //   port: 3000,
    // },
})

**Comment by dgozman** [comment]
@dipesh17 In general, you are going in the right direction. However, a few issues:
- `ccBaseURL` is never defined in your snippet, I guess it's coming from somewhere else?
- `addSegmentCookies()` returns `Promise`, so you have to `await` it before spreading.
- Config file does not support any async operations, like the `context.get()` you are calling. You should move the token retrieval to global setup, dependency project, beforeAll hook or a worker fixture. Take a look at the [auth guide](https://playwright.dev/docs/auth) that solves a similar problem.

Let me know if that helps.

**Comment by dgozman** [comment]
Closing as per above. If you still encounter problems, please file a new issue with a repro and link to this one.
